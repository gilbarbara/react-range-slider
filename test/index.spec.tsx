import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import RangeSlider, { RangeSliderProps } from '../src';

const mockOnAfterEnd = jest.fn();
const mockOnChange = jest.fn();
const mockOnDragEnd = jest.fn();

function setup(axis = 'x') {
  return render(
    <RangeSlider
      axis={axis as RangeSliderProps['axis']}
      className="test"
      onAfterEnd={mockOnAfterEnd}
      onChange={mockOnChange}
      onDragEnd={mockOnDragEnd}
    />,
  );
}

describe('RangeSlider', () => {
  beforeAll(() => {
    // @ts-ignore
    Element.prototype.getBoundingClientRect = () => ({
      bottom: 0,
      height: 20,
      left: 50,
      right: 0,
      top: 50,
      width: 200,
    });
  });

  afterEach(() => {
    mockOnAfterEnd.mockClear();
    mockOnChange.mockClear();
    mockOnDragEnd.mockClear();
  });

  describe('with `x` axis', () => {
    it('should render properly', () => {
      const { container } = setup();

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should handle mouse events', () => {
      setup();

      fireEvent.mouseDown(screen.getByRole('slider'), {
        clientX: 100,
        clientY: 0,
        currentTarget: {},
      });

      fireEvent.mouseMove(document, {
        clientX: 230,
        clientY: 0,
        currentTarget: {},
      });

      fireEvent.mouseUp(document, {
        clientX: 230,
        clientY: 0,
        currentTarget: {},
      });

      expect(mockOnAfterEnd).toHaveBeenLastCalledWith({ x: 65, y: 0 }, expect.any(Object));
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 65, y: 0 }, expect.any(Object));
      expect(mockOnDragEnd).toHaveBeenLastCalledWith({ x: 65, y: 0 }, expect.any(Object));
      expect(screen.getByRole('slider')).toMatchSnapshot();
    });

    it('should handle touch events', () => {
      setup();

      fireEvent.touchStart(screen.getByRole('slider'), {
        touches: [
          {
            clientX: 100,
            clientY: 0,
          },
        ],
        currentTarget: {},
      });

      fireEvent.touchMove(document, {
        touches: [
          {
            clientX: 150,
            clientY: 0,
          },
        ],
        currentTarget: {},
      });

      fireEvent.touchEnd(document, {
        touches: [],
        currentTarget: {},
      });

      expect(mockOnAfterEnd).toHaveBeenLastCalledWith({ x: 25, y: 0 }, expect.any(Object));
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 25, y: 0 }, expect.any(Object));
      expect(mockOnDragEnd).toHaveBeenLastCalledWith({ x: 25, y: 0 }, expect.any(Object));
      expect(screen.getByRole('slider')).toMatchSnapshot();
    });

    it('should handle clicks on the track', () => {
      setup();
      const [track] = screen.getAllByRole('presentation');

      fireEvent.click(track, {
        clientX: 80,
        clientY: 0,
        currentTarget: {},
      });

      expect(mockOnAfterEnd).toHaveBeenLastCalledWith({ x: 15, y: 0 }, expect.any(Object));
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 15, y: 0 }, expect.any(Object));
      expect(screen.getByRole('slider')).toMatchSnapshot();
    });

    it('should handle focus, keydown and blur', () => {
      setup();

      fireEvent.focus(screen.getByRole('slider'));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowRight' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 1, y: 0 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowUp' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 2, y: 0 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowLeft' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 1, y: 0 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowDown' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 0 }, expect.any(Object));

      fireEvent.blur(screen.getByRole('slider'));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowUp' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 0 }, expect.any(Object));
    });
  });

  describe('with `xy` axis', () => {
    it('should render properly', () => {
      const { container } = setup('xy');

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should handle focus, keydown and blur', () => {
      setup('xy');

      fireEvent.focus(screen.getByRole('slider'));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowRight' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 1, y: 0 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowUp' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 1, y: 1 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowLeft' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 1 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowDown' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 0 }, expect.any(Object));

      fireEvent.blur(screen.getByRole('slider'));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowUp' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 0 }, expect.any(Object));
    });
  });

  describe('with `y` axis', () => {
    it('should render properly', () => {
      const { container } = setup('y');

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should handle focus, keydown and blur', () => {
      setup('y');

      fireEvent.focus(screen.getByRole('slider'));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowRight' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 1 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowUp' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 2 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowLeft' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 1 }, expect.any(Object));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowDown' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 0 }, expect.any(Object));

      fireEvent.blur(screen.getByRole('slider'));

      fireEvent.keyDown(screen.getByRole('slider'), { code: 'ArrowUp' });
      expect(mockOnChange).toHaveBeenLastCalledWith({ x: 0, y: 0 }, expect.any(Object));
    });
  });
});
