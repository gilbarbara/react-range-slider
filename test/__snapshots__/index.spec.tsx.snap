// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider with \`x\` axis should handle clicks on the track 1`] = `
<span
  aria-label="slider handle"
  aria-orientation="horizontal"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="15"
  class="test__thumb"
  role="slider"
  style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; height: 26px; left: -5px; top: -3px; width: 10px;"
  tabindex="0"
/>
`;

exports[`RangeSlider with \`x\` axis should handle mouse events 1`] = `
<span
  aria-label="slider handle"
  aria-orientation="horizontal"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="65"
  class="test__thumb"
  role="slider"
  style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; height: 26px; left: -5px; top: -3px; width: 10px;"
  tabindex="0"
/>
`;

exports[`RangeSlider with \`x\` axis should handle touch events 1`] = `
<span
  aria-label="slider handle"
  aria-orientation="horizontal"
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="25"
  class="test__thumb"
  role="slider"
  style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; height: 26px; left: -5px; top: -3px; width: 10px;"
  tabindex="0"
/>
`;

exports[`RangeSlider with \`x\` axis should render properly 1`] = `
<div
  class="test"
  style="box-sizing: border-box; display: inline-block; padding: 6px; transition: height 0.4s, width 0.4s; height: 32px; width: 100%;"
>
  <div
    class="test__track"
    role="presentation"
    style="background-color: rgb(204, 204, 204); border-radius: 3px; box-sizing: border-box; height: 20px; position: relative; width: 100%;"
  >
    <div
      class="test__range"
      style="width: 0%; background-color: rgb(0, 123, 255); border-radius: 3px; position: absolute; height: 100%; top: 0px;"
    />
    <div
      role="presentation"
      style="box-sizing: border-box; height: 20px; position: absolute; transition: height 0.4s, width 0.4s; width: 20px; left: 0%; bottom: 0%;"
    >
      <span
        aria-label="slider handle"
        aria-orientation="horizontal"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="test__thumb"
        role="slider"
        style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; height: 26px; left: -5px; top: -3px; width: 10px;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`RangeSlider with \`xy\` axis should render properly 1`] = `
<div
  class="test"
  style="box-sizing: border-box; display: inline-block; padding: 6px; transition: height 0.4s, width 0.4s; height: 100%; width: 100%;"
>
  <div
    class="test__track"
    role="presentation"
    style="background-color: rgb(204, 204, 204); border-radius: 3px; box-sizing: border-box; height: 100%; position: relative; width: 100%; min-height: 50px;"
  >
    <div
      class="test__range"
      style="height: 0%; width: 0%; background-color: rgb(0, 123, 255); border-radius: 3px; position: absolute; bottom: 0px;"
    />
    <div
      role="presentation"
      style="box-sizing: border-box; height: 20px; position: absolute; transition: height 0.4s, width 0.4s; width: 20px; left: 0%; bottom: 0%;"
    >
      <span
        aria-label="slider handle"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="test__thumb"
        role="slider"
        style="background-color: transparent; border: 2px solid #000; border-radius: 50%; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; bottom: -10px; height: 20px; left: -10px; width: 20px;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;

exports[`RangeSlider with \`y\` axis should render properly 1`] = `
<div
  class="test"
  style="box-sizing: border-box; display: inline-block; padding: 6px; transition: height 0.4s, width 0.4s; height: 100%; width: 32px;"
>
  <div
    class="test__track"
    role="presentation"
    style="background-color: rgb(204, 204, 204); border-radius: 3px; box-sizing: border-box; height: 100%; position: relative; width: 20px; min-height: 50px;"
  >
    <div
      class="test__range"
      style="height: 0%; background-color: rgb(0, 123, 255); border-radius: 3px; position: absolute; bottom: 0px; left: 0px; width: 100%;"
    />
    <div
      role="presentation"
      style="box-sizing: border-box; height: 20px; position: absolute; transition: height 0.4s, width 0.4s; width: 20px; left: 0%; bottom: 0%;"
    >
      <span
        aria-label="slider handle"
        aria-orientation="vertical"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="test__thumb"
        role="slider"
        style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; bottom: -5px; height: 10px; left: -3px; width: 26px;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
